File: app/src/main/java/com/fonfon/noloss/ui/main/DevicesAdapter.java
package com.fonfon.noloss.ui.main;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.fonfon.noloss.R;
import com.fonfon.noloss.lib.Device;
import com.jakewharton.rxbinding2.view.RxView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

final class DevicesAdapter extends RecyclerView.Adapter<DevicesAdapter.Holder> {

  private final ArrayList<Device> devices = new ArrayList<>();
  private final DeviceAdapterListener listener;

  private final int red;
  private final int green;
  private final String connected;
  private final String disconnected;

  DevicesAdapter(Context context, DeviceAdapterListener listener) {
    this.listener = listener;
    red = ContextCompat.getColor(context, R.color.mojo);
    green = ContextCompat.getColor(context, R.color.fern);

    connected = context.getString(R.string.status_connected);
    disconnected = context.getString(R.string.status_disconnected);
  }

  @Override
  public DevicesAdapter.Holder onCreateViewHolder(ViewGroup parent, int viewType) {
    return new Holder(
        LayoutInflater.from(parent.getContext()).inflate(R.layout.item_device, parent, false)
    );
  }

  @Override
  public void onBindViewHolder(final DevicesAdapter.Holder holder, int position) {
    Device device = devices.get(position);

    holder.image.setImageBitmap(Device.getBitmapImage(device.getImage(), holder.image.getContext().getResources()));

    int tintColor = device.isConnected() ? Color.BLACK : Color.WHITE;
    holder.toolbar.setTitle(device.getName());
    holder.toolbar.setTitleTextColor(tintColor);

    holder.toolbar.setSubtitle(device.getAddress());
    holder.toolbar.setSubtitleTextColor(tintColor);

    holder.toolbar.setBackgroundColor(device.isConnected() ? green : red);

    Drawable moreIcon = ContextCompat.getDrawable(holder.toolbar.getContext(), R.drawable.ic_more);
    DrawableCompat.setTint(moreIcon, tintColor);
    holder.toolbar.setOverflowIcon(moreIcon);

    holder.viewStatus.setEnabled(device.isConnected());
    holder.fabAlert.setImageResource(device.isAlerted() ? R.drawable.ic_volume_off : R.drawable.ic_volume_up);
    holder.textStatus.setText(device.isConnected() ? connected : disconnected);
  }

  @Override
  public int getItemCount() {
    return devices.size();
  }

  void setDevices(@NonNull List<Device> devices) {
    this.devices.clear();
    this.devices.addAll(devices);

    notifyDataSetChanged();
  }

  class Holder extends RecyclerView.ViewHolder {

    @BindView(R.id.imageDevice)
    ImageView image;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.viewStatus)
    TextView viewStatus;
    @BindView(R.id.textStatus)
    TextView textStatus;
    @BindView(R.id.fabAlert)
    FloatingActionButton fabAlert;

    Holder(View view) {
      super(view);
      ButterKnife.bind(this, view);

      toolbar.inflateMenu(R.menu.device_detail);
      fabAlert.setOnClickListener(v -> listener.onAlert(devices.get(getAdapterPosition())));

      toolbar.setOnMenuItemClickListener(item -> {
        switch (item.getItemId()) {
          case R.id.menu_delete:
            listener.onDelete(devices.get(getAdapterPosition()));
            break;
          case R.id.menu_rename:
            listener.onRename(devices.get(getAdapterPosition()));
            break;
          case R.id.menu_image:
            listener.onEditImage(devices.get(getAdapterPosition()));
        }
        return false;
      });
    }
  }

  interface DeviceAdapterListener {
    void onRename(Device device);

    void onEditImage(Device device);

    void onDelete(Device device);

    void onAlert(Device device);
  }
}

File: app/src/main/java/com/fonfon/noloss/ui/main/DevicesAdapter.kt
package com.fonfon.noloss.ui.main

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.fonfon.noloss.R
import com.fonfon.noloss.lib.Device
import kotlinx.android.synthetic.main.item_device.view.*
import java.util.*

internal class DevicesAdapter(context: Context, private val listener: DeviceAdapterListener) : RecyclerView.Adapter<DevicesAdapter.Holder>() {

  private val devices = ArrayList<Device>()

  private val connected = context.getString(R.string.status_connected)
  private val disconnected = context.getString(R.string.status_disconnected)

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = Holder(
      LayoutInflater.from(parent.context).inflate(R.layout.item_device, parent, false)
  )

  override fun onBindViewHolder(holder: DevicesAdapter.Holder, position: Int) {
    val device = devices[position]

    holder.itemView.toolbar.title = device.name
    holder.itemView.toolbar.subtitle = device.address

    holder.itemView.viewStatus.isEnabled = device.isConnected
    holder.itemView.fabAlert.setImageResource(if (device.isAlerted) R.drawable.ic_volume_off else R.drawable.ic_volume_up)
    holder.itemView.textStatus.text = if (device.isConnected) connected else disconnected
  }

  override fun getItemCount() = devices.size

  fun setDevices(devices: List<Device>) {
    this.devices.clear()
    this.devices.addAll(devices)

    notifyDataSetChanged()
  }

  internal inner class Holder(view: View) : RecyclerView.ViewHolder(view) {

    init {
      itemView.toolbar.inflateMenu(R.menu.device_detail)
      itemView.fabAlert.setOnClickListener { listener.onAlert(devices[adapterPosition]) }
      itemView.toolbar.setOnMenuItemClickListener { item ->
        when (item.itemId) {
          R.id.menu_delete -> listener.onDelete(devices[adapterPosition])
          R.id.menu_rename -> listener.onRename(devices[adapterPosition])
        }
        false
      }
    }
  }

  internal interface DeviceAdapterListener {
    fun onRename(device: Device)

    fun onDelete(device: Device)

    fun onAlert(device: Device)
  }
}
